### PR TITLE
Introduce InternalWipe

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

> /kind api-change
/kind bug
> /kind ci
> /kind cleanup
> /kind dependency-change
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake
> /kind other

#### What this PR does / why we need it:
On node reboot, there are situations where CNI plugins aren't able to cleanup the resources they acquire. They rely on the caller of CNI add to send a corresponding CNI del.

The problem is, CRI-O was previously not able to send the CNI del, because `crio wipe` removed the evidence of the container's (and pod's) existence. As such, on every reboot, some CNI plugins have the ability to leak resources.

This PR is the first step in remediating it. It moves the responsibility of wiping images and containers from `crio wipe` to server startup. This behavior is gated by a config value `internal_wipe`.

It leaves the responsibility of checking whether the node shutdown and was able to sync to `crio wipe`, as well as leaves `crio wipe -f` which has proven useful to wipe the state of a node.

We also adapt the tests to the new behavior (previously, we were wiping the pause container, which we no longer do. I think this is the correct behavior)

#### Which issue(s) this PR fixes:
fixes half of https://github.com/cri-o/cri-o/issues/4727

#### Special notes for your reviewer:
built on top of https://github.com/cri-o/cri-o/pull/4763
#### Does this PR introduce a user-facing change?

```release-note
Add the config field `internal_wipe` which moves the responsibility of wiping containers after a reboot and images after an upgrade from the external binary `crio wipe` to the main crio server. This has a handful of advantages, the main one being crio is now better able to cleanup CNI resources after a reboot.
```

### CODE REVIEW

File: docs/crio.8.md
Comment (TomSweeneyRedHat) [inline docs/crio.8.md:236]: what does "cleanup the containers" mean?  I don't see another instance of "cleanup" in this entire man page.  A bit more detail please.  It would also be good to be specific about what wipe does.

File: docs/crio.conf.5.md
Comment (TomSweeneyRedHat) [inline docs/crio.conf.5.md:59]: see prior

File: internal/resourcestore/resourcecleaner.go
Comment (haircommander) [inline internal/resourcestore/resourcecleaner.go:29]: I don't think we're supposed to save ctx as a member variable

**Comment by haircommander** [comment]
ref BZ https://bugzilla.redhat.com/show_bug.cgi?id=1948047 https://bugzilla.redhat.com/show_bug.cgi?id=1948137

File: internal/resourcestore/resourcecleaner.go
Comment (saschagrunert) [reply internal/resourcestore/resourcecleaner.go:29]: The main reason is to be able to use `log.Errof` in `Cleanup()`. Do you think we should pass the context on `Cleanup(ctx)` as alternative?

**Comment by saschagrunert** [comment]
@haircommander added another commit to run the cleanup in parallel

**Comment by saschagrunert** [comment]
Something in the pod sandbox removal it broken since ec883ab13d310761651a120858763d1f64bf57d0 which made the integration tests fail.

**Comment by saschagrunert** [comment]
Bundle test is not broken by this PR.

**Comment by saschagrunert** [comment]
Some integration tests are still broken due to the sandbox cleanup behavior.

**Comment by saschagrunert** [comment]
It looks like that the CNI del for the bridge plugin (v0.9.1) does fail if the network namespace (`/var/run/netns/…`) does not exist. In that case it will also not cleanup the resources in `/var/lib/cni/results`. I'm not sure if we can do anything from the CRI-O side to solve this kind of issue.

**Comment by saschagrunert** [comment]
I think I found the issue in the parallel cleanup: We have to cleanup the namespace resources always after network teardown. This may solve the error from the CNI plugin.

**Comment by saschagrunert** [comment]
@haircommander please take a look. I reverted some of my experimental commits and it seems that still some integration tests are failing due to the parallel cleanup.

File: server/sandbox_run_linux.go
Comment (saschagrunert) [inline server/sandbox_run_linux.go:725]: The issue with that is now if we return an error, it will cause a retry of the function. I don't think that it can work this way.

**Comment by mrunalp** [comment]
@kolyshkin @QiWang19 ptal

**Review by saschagrunert — APPROVED** [review]
LGTM

**Review by QiWang19 — COMMENTED** [review]
/lgtm

**Comment by haircommander** [comment]
/hold

the unit test failure seems to be legit. I think I messed something up...

**Comment by haircommander** [comment]
/hold cancel
fixed
/retest

File: internal/lib/container_server.go
Comment (kolyshkin) [inline internal/lib/container_server.go:223]: Do you mean
```suggestion
			if nsErr := namespaceToJoin.joinFunc(path); nsErr != nil {
```

Comment (kolyshkin) [reply internal/lib/container_server.go:223]: Oh I see this error is in the current codebase. Makes sense to have a separate PR to fix it.

Comment (haircommander) [reply internal/lib/container_server.go:223]: ach damn

Comment (haircommander) [reply internal/lib/container_server.go:223]: https://github.com/cri-o/cri-o/pull/4796 to fix, I'll rebase this on that when evaluating review comments

File: cmd/crio/wipe.go
Comment (kolyshkin) [inline cmd/crio/wipe.go:71]: Is there some punctuation missing between the first and the second line? It's a bit hard to parse this comment.

File: server/server.go
Comment (kolyshkin) [inline server/server.go:504]: perhaps `log.Warnf(ctx, "error while checking ....`?

File: server/server.go
Comment (kolyshkin) [inline server/server.go:514]: `log.Warnf`?

File: server/server.go
Comment (kolyshkin) [inline server/server.go:521]: `var imagesToWipe []string`

(`make` makes sense if we know the size, otherwise it just makes things look more complicated when they should)

File: server/server.go
Comment (kolyshkin) [inline server/server.go]: What is the point of using the cleanupFunc here? It looks like it's equivalent to this (simpler) code:

```go
for _, sb := range s.ContainerServer.ListSandboxes() {
   err := s.stopPodSandbox(ctx, sb)
   if err == nil {
      err = s.removePodSandbox(ctx, sb)
   }
   if err != nil {
      log.Warnf(ctx, "Failed to cleanup pod %s: %v", sb.ID(), err)
   }
}
```

File: test/crio-wipe.bats
Comment (kolyshkin) [inline test/crio-wipe.bats:64]: I doubt that `export` is needed here. Is this just to eliminate a shellcheck warning?

File: internal/resourcestore/resourcecleaner.go
Comment (kolyshkin) [reply internal/resourcestore/resourcecleaner.go]: Yes, I think it's better to pass `ctx` to `Cleanup` directly.

File: internal/resourcestore/resourcecleaner.go
Comment (kolyshkin) [inline internal/resourcestore/resourcecleaner.go]: why not
```suggestion
		if err := retry(r.funcs[i]); err != nil {
```
?

File: internal/resourcestore/resourcecleaner.go
Comment (kolyshkin) [inline internal/resourcestore/resourcecleaner.go:74]: is there any way not to discard the real error returned from `fn()`?

File: server/sandbox_network.go
Comment (kolyshkin) [inline server/sandbox_network.go:188]: No need to explicitly initialize to `0` -- this is the default value.

File: server/server.go
Comment (kolyshkin) [inline server/server.go:543]:
```suggestion
				log.Warnf(ctx, "Failed to cleanup pod %s (will retry): %v", sb.ID(), err)
```
and maybe demote from warning to debug?

File: test/crio-wipe.bats
Comment (kolyshkin) [inline test/crio-wipe.bats:268]: not sure if `export` is needed here.

File: test/crio-wipe.bats
Comment (kolyshkin) [inline test/crio-wipe.bats:273]: ditto

File: test/crio-wipe.bats
Comment (kolyshkin) [inline test/crio-wipe.bats:302]: easier to just do
```suggestion
	! ls "$CNI_RESULTS_DIR"/*"$pod_id"*
```

File: internal/resourcestore/resourcecleaner.go
Comment (kolyshkin) [inline internal/resourcestore/resourcecleaner.go]: An earlier commit from this PR adds ctx to `ResourceCleaner`, and now you remove it. It's better to fix the first commit right away (less changes in history, easier to review).

File: internal/storage/runtime.go
Comment (kolyshkin) [inline internal/storage/runtime.go]: This function was removed by an earlier commit, and now it's being re-added. I don't follow...

File: server/sandbox_run_linux.go
Comment (kolyshkin) [inline server/sandbox_run_linux.go:725]: typo: `stoppign`

File: server/sandbox_run_linux.go
Comment (kolyshkin) [inline server/sandbox_run_linux.go:721]: please remove

File: server/server.go
Comment (haircommander) [reply server/server.go]: because we expect this to fail in some cases (if the node has restarted but the networking plugin pods haven't been created, then any non-host network pods can't have their networking resources cleaned up)

File: test/crio-wipe.bats
Comment (haircommander) [reply test/crio-wipe.bats:64]: I think export is needed because this field is used outside this function

File: server/server.go
Comment (haircommander) [reply server/server.go]: oh I see, we don't yet actually do that. I think it was just to reduce churn between commits, but I can drop it for this commit and update in a later one

File: test/crio-wipe.bats
Comment (haircommander) [reply test/crio-wipe.bats:64]: hm not needed :)

File: internal/storage/runtime.go
Comment (haircommander) [reply internal/storage/runtime.go]: I have reworked the commits to only take the lint fixing parts of this commit. I believe this was readded by @saschagrunert to debug failures that other commits ended up fixing. we'll see how that works...

File: docs/crio.8.md
Comment (haircommander) [reply docs/crio.8.md:236]: alright I've updated the verbiage to hopefully be more clear

File: internal/resourcestore/resourcecleaner.go
Comment (mrunalp) [reply internal/resourcestore/resourcecleaner.go:74]: Can we log the error here?

File: internal/resourcestore/resourcecleaner.go
Comment (mrunalp) [inline internal/resourcestore/resourcecleaner.go]: We should bubble up for logging up the stack here.

Comment (haircommander) [reply internal/resourcestore/resourcecleaner.go]: I have updated it to always return the error here

File: internal/resourcestore/resourcecleaner.go
Comment (haircommander) [reply internal/resourcestore/resourcecleaner.go:75]: alright, done

**Comment by haircommander** [comment]
okay comments addressed and I think this is ready again
@kolyshkin @saschagrunert @mrunalp @umohnani8 @QiWang19 PTAL

**Comment by saschagrunert** [comment]
@haircommander this needs a rebase

**Pull request merged**
